### PR TITLE
New Expression: Free / Max / Total Memory

### PR DESCRIPTION
**Pull request by DelayedGaming**
### Description
This PR adds an expression to get the free, max, or total memory of the server to Skript!

---
**Target Minecraft Versions:** any
**Requirements:** -
**Related Issues:** #5628

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprMemory.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	}
+
+	@Override
+	public Class<? extends Number> getReturnType() {
+		return Number.class;
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return (type == Type.FREE ? "free" : type == Type.MAXIMUM ? "maximum" : "total") + " memory";
```
Comment (UnderscoreTud) [inline src/main/java/ch/njol/skript/expressions/ExprMemory.java:91]:
```suggestion
		return type.name().toLowercase(Locale.ENGLISH) + " memory";
```

**Review by UnderscoreTud — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprMemory.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		} else if (parseResult.hasTag("max")) {
+			type = Type.MAXIMUM;
+		} else {
+			type = Type.TOTAL;
+		}
+		return true;
+	}
+
+	@Override
+	protected @Nullable Number[] get(Event event) {
```
[the remark this answers is not in the document]
Comment (DelayedGaming) [reply src/main/java/ch/njol/skript/expressions/ExprMemory.java:65]: Removed `@Nullable` annotation as not possible of returning an empty array in this expression.

File: src/main/java/ch/njol/skript/expressions/ExprMemory.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+			case FREE:
+				memory = RUNTIME.freeMemory();
+				break;
+			case MAXIMUM:
+				memory = RUNTIME.maxMemory();
+				break;
+			case TOTAL:
+				memory = RUNTIME.totalMemory();
+		}
+		return CollectionUtils.array(memory.longValue() * 1.024e-6);
```
[the remark this answers is not in the document]
Comment (DelayedGaming) [reply src/main/java/ch/njol/skript/expressions/ExprMemory.java]: Awaiting results from a pending debation in skUnity's SkriptLang channel.

[the remark this answers is not in the document]
Comment (DelayedGaming) [reply src/main/java/ch/njol/skript/expressions/ExprMemory.java]: Using 1E-6 in a constant field.

File: src/main/java/ch/njol/skript/expressions/ExprMemory.java [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+		"\twait 5 ticks"
+})
+@Since("INSERT VERSION")
+public class ExprMemory extends SimpleExpression<Double> {
+
+	private static final double BYTES_IN_MEGABYTES = 1E-6;
+	private static final Runtime RUNTIME = Runtime.getRuntime();
+
+	static {
+		Skript.registerExpression(ExprMemory.class, Double.class, ExpressionType.SIMPLE, "[the] (:free|:max[imum]|total) (memory|ram)");
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprMemory.java:51]:
```suggestion
		Skript.registerExpression(ExprMemory.class, Double.class, ExpressionType.SIMPLE, "[the] [server] (:free|:max[imum]|total) (memory|ram)");
```

File: src/main/java/ch/njol/skript/expressions/ExprMemory.java [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+	public Class<? extends Double> getReturnType() {
+		return Double.class;
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return type.name().toLowerCase(Locale.ENGLISH) + " memory";
+	}
+
+	private enum Type {
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprMemory.java:100]: Let's move enum above the field declaration L54

**Review by AyhamAl-Ali — APPROVED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprMemory.java [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+		"\twait 5 ticks"
+})
+@Since("INSERT VERSION")
+public class ExprMemory extends SimpleExpression<Double> {
+
+	private static final double BYTES_IN_MEGABYTES = 1E-6;
+	private static final Runtime RUNTIME = Runtime.getRuntime();
+
+	static {
+		Skript.registerExpression(ExprMemory.class, Double.class, ExpressionType.SIMPLE, "[the] [server] (:free|:max[imum]|total) (memory|ram)");
```
Comment (AyhamAl-Ali) [inline src/main/java/ch/njol/skript/expressions/ExprMemory.java:51]: If the user wrote `maximum` the tag would be `maximum` not `max`
```suggestion
		Skript.registerExpression(ExprMemory.class, Double.class, ExpressionType.SIMPLE, "[the] [server] (:free|max:max[imum]|total) (memory|ram)");
```

**Pull request merged**